File: lib/components/cardlist.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import '../theme/theme.dart';

const double kGroupedOuterRadius = 12;
const double kGroupedInnerRadius = 4;
const double kGroupedGap = 3;
const double kGroupedHorizontalPadding = 14;
const EdgeInsets kGroupedCardPadding = EdgeInsets.fromLTRB(12, 6, 8, 6);
const double kGroupedGridMaxExtent = 104;
const double kGroupedGridTileHeight = 104;
const double kGroupedTitleSize = 12.5;
const FontWeight kGroupedTitleWeight = FontWeight.w600;
const EdgeInsets kGroupedTitlePadding = EdgeInsets.fromLTRB(12, 2, 12, 6);

class GroupedCardCorners extends InheritedWidget {
  const GroupedCardCorners({
    super.key,
    required this.radius,
    required super.child,
  });

  final BorderRadius radius;

  static const outer = Radius.circular(kGroupedOuterRadius);
  static const inner = Radius.circular(kGroupedInnerRadius);

  static BorderRadius of(BuildContext context) =>
      context
          .dependOnInheritedWidgetOfExactType<GroupedCardCorners>()
          ?.radius ??
      const BorderRadius.all(outer);

  @override
  bool updateShouldNotify(GroupedCardCorners oldWidget) =>
      radius != oldWidget.radius;
}

class GroupedCard extends StatelessWidget {
  const GroupedCard({
    super.key,
    this.onTap,
    this.padding = kGroupedCardPadding,
    this.background,
    required this.child,
  });

  final VoidCallback? onTap;
  final EdgeInsetsGeometry padding;
  final Widget? background;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    final content = InkWell(
      onTap: onTap,
      child: Padding(padding: padding, child: child),
    );
    return Material(
      color: colors.card,
      borderRadius: GroupedCardCorners.of(context),
      clipBehavior: Clip.antiAlias,
      child: background == null
          ? content
          : Stack(children: [background!, content]),
    );
  }
}

Widget _groupTitle(BuildContext context, String? title, Widget? header) {
  if (header != null) return header;
  if (title == null) return const SizedBox.shrink();
  return Padding(
    padding: kGroupedTitlePadding,
    child: Text(
      title,
      style: TextStyle(
        fontSize: kGroupedTitleSize,
        fontWeight: kGroupedTitleWeight,
        color: context.appColors.textSecondary,
      ),
    ),
  );
}

class GroupedCardList<T> extends StatelessWidget {
  const GroupedCardList({
    super.key,
    this.title,
    this.header,
    required this.items,
    required this.itemBuilder,
    this.onTap,
    this.cardPadding = kGroupedCardPadding,
    this.backgroundBuilder,
    this.wrapItems = true,
  });

  final String? title;
  final Widget? header;
  final List<T> items;
  final Widget Function(BuildContext context, T item) itemBuilder;
  final VoidCallback? Function(T item)? onTap;
  final EdgeInsetsGeometry cardPadding;
  final Widget? Function(T item)? backgroundBuilder;

  final bool wrapItems;

  @override
  Widget build(BuildContext context) {
    final last = items.length - 1;
    return Padding(
      padding: const EdgeInsets.symmetric(
        horizontal: kGroupedHorizontalPadding,
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          if (header != null || title != null)
            _groupTitle(context, title, header),
          for (var i = 0; i < items.length; i++) ...[
            if (i > 0) const SizedBox(height: kGroupedGap),
            GroupedCardCorners(
              radius: BorderRadius.vertical(
                top: i == 0
                    ? GroupedCardCorners.outer
                    : GroupedCardCorners.inner,
                bottom: i == last
                    ? GroupedCardCorners.outer
                    : GroupedCardCorners.inner,
              ),
              child: wrapItems
                  ? GroupedCard(
                      onTap: onTap?.call(items[i]),
                      padding: cardPadding,
                      background: backgroundBuilder?.call(items[i]),
                      child: itemBuilder(context, items[i]),
                    )
                  : itemBuilder(context, items[i]),
            ),
          ],
        ],
      ),
    );
  }
}

class GroupedCardGrid<T> extends StatelessWidget {
  const GroupedCardGrid({
    super.key,
    this.title,
    this.header,
    required this.items,
    required this.itemBuilder,
    this.onTap,
    this.backgroundBuilder,
    this.cardPadding = kGroupedCardPadding,
    this.maxCrossAxisExtent = kGroupedGridMaxExtent,
    this.mainAxisExtent = kGroupedGridTileHeight,
    this.crossAxisCount,
    this.spacing = kGroupedGap,
    this.wrapItems = true,
  });

  final String? title;
  final Widget? header;
  final List<T> items;
  final Widget Function(BuildContext context, T item) itemBuilder;
  final VoidCallback? Function(T item)? onTap;
  final Widget? Function(T item)? backgroundBuilder;
  final EdgeInsetsGeometry cardPadding;
  final double maxCrossAxisExtent;

  /// Fixed cell height; when null the row sizes to the tallest cell.
  final double? mainAxisExtent;

  /// Forces the column count, ignoring [maxCrossAxisExtent] when set.
  final int? crossAxisCount;
  final double spacing;
  final bool wrapItems;

  Widget _cell(BuildContext context, T item) {
    if (!wrapItems) return itemBuilder(context, item);
    return GroupedCard(
      onTap: onTap?.call(item),
      padding: cardPadding,
      background: backgroundBuilder?.call(item),
      child: itemBuilder(context, item),
    );
  }

  BorderRadius _cellRadius(int index, int columns, int total) {
    final col = index % columns;
    final top = index - columns < 0;
    final bottom = index + columns >= total;
    final left = col == 0;
    final right = col == columns - 1 || index + 1 >= total;
    Radius corner(bool a, bool b) =>
        (a && b) ? GroupedCardCorners.outer : GroupedCardCorners.inner;
    return BorderRadius.only(
      topLeft: corner(top, left),
      topRight: corner(top, right),
      bottomLeft: corner(bottom, left),
      bottomRight: corner(bottom, right),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(
        horizontal: kGroupedHorizontalPadding,
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          if (header != null || title != null)
            _groupTitle(context, title, header),
          LayoutBuilder(
            builder: (context, constraints) {
              final width = constraints.maxWidth;
              final columns =
                  crossAxisCount ??
                  math.max(1, (width / (maxCrossAxisExtent + spacing)).ceil());
              final fixedHeight = mainAxisExtent;
              final rows = <Widget>[];
              for (var start = 0; start < items.length; start += columns) {
                final cells = <Widget>[];
                for (var c = 0; c < columns; c++) {
                  if (c > 0) cells.add(SizedBox(width: spacing));
                  final index = start + c;
                  Widget cell;
                  if (index < items.length) {
                    cell = GroupedCardCorners(
                      radius: _cellRadius(index, columns, items.length),
                      child: _cell(context, items[index]),
                    );
                    if (fixedHeight != null) {
                      cell = SizedBox(height: fixedHeight, child: cell);
                    }
                  } else {
                    cell = const SizedBox.shrink();
                  }
                  cells.add(Expanded(child: cell));
                }
                if (rows.isNotEmpty) rows.add(SizedBox(height: spacing));
                final row = Row(
                  crossAxisAlignment: fixedHeight == null
                      ? CrossAxisAlignment.stretch
                      : CrossAxisAlignment.center,
                  children: cells,
                );
                rows.add(
                  fixedHeight == null ? IntrinsicHeight(child: row) : row,
                );
              }
              return Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: rows,
              );
            },
          ),
        ],
      ),
    );
  }
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/components/CardList.kt
package com.example.ohmyssh.components

import androidx.compose.foundation.background
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.clickable
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.ohmyssh.theme.appColors
import kotlin.math.ceil
import kotlin.math.max

val kGroupedOuterRadius = 12.dp
val kGroupedInnerRadius = 4.dp
val kGroupedGap = 3.dp
val kGroupedHorizontalPadding = 14.dp
val kGroupedCardPadding = PaddingValues(start = 12.dp, top = 6.dp, end = 8.dp, bottom = 6.dp)
val kGroupedGridMaxExtent = 104.dp
val kGroupedGridTileHeight = 104.dp

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun GroupedCard(
    shape: RoundedCornerShape,
    onTap: (() -> Unit)? = null,
    onLongPress: (() -> Unit)? = null,
    padding: PaddingValues = kGroupedCardPadding,
    background: (@Composable () -> Unit)? = null,
    contentAlignment: Alignment = Alignment.CenterStart,
    modifier: Modifier = Modifier,
    gestureModifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val colors = appColors
    Box(
        modifier
            .clip(shape)
            .background(colors.card)
            .let { base ->
                when {
                    // One gesture owner for the whole card: a detectTapGestures
                    // inside a row would consume the press and the card would
                    // stop reacting to taps entirely.
                    onTap != null && onLongPress != null -> base.combinedClickable(
                        onClick = onTap,
                        onLongClick = onLongPress,
                    )
                    onTap != null -> base.clickable(onClick = onTap)
                    onLongPress != null -> base.combinedClickable(
                        onClick = {},
                        onLongClick = onLongPress,
                    )
                    else -> base
                }
            }
            .then(gestureModifier),
    ) {
        background?.invoke()
        Box(
            Modifier.fillMaxSize().padding(padding),
            contentAlignment = contentAlignment,
        ) {
            content()
        }
    }
}

@Composable
private fun GroupTitle(title: String?, header: (@Composable () -> Unit)?) {
    if (header != null) {
        header()
        return
    }
    if (title == null) return
    Text(
        title,
        style = TextStyle(
            fontSize = 12.5.sp,
            fontWeight = FontWeight.W600,
            color = appColors.textSecondary,
        ),
        modifier = Modifier.padding(start = 12.dp, top = 2.dp, end = 12.dp, bottom = 6.dp),
    )
}

fun groupedListShape(index: Int, count: Int): RoundedCornerShape = RoundedCornerShape(
    topStart = if (index == 0) kGroupedOuterRadius else kGroupedInnerRadius,
    topEnd = if (index == 0) kGroupedOuterRadius else kGroupedInnerRadius,
    bottomStart = if (index == count - 1) kGroupedOuterRadius else kGroupedInnerRadius,
    bottomEnd = if (index == count - 1) kGroupedOuterRadius else kGroupedInnerRadius,
)

fun groupedRowShape(index: Int, count: Int): RoundedCornerShape = RoundedCornerShape(
    topStart = if (index == 0) kGroupedOuterRadius else kGroupedInnerRadius,
    bottomStart = if (index == 0) kGroupedOuterRadius else kGroupedInnerRadius,
    topEnd = if (index == count - 1) kGroupedOuterRadius else kGroupedInnerRadius,
    bottomEnd = if (index == count - 1) kGroupedOuterRadius else kGroupedInnerRadius,
)

private fun listShape(index: Int, last: Int): RoundedCornerShape =
    groupedListShape(index, last + 1)

@Composable
fun <T> GroupedCardList(
    items: List<T>,
    title: String? = null,
    header: (@Composable () -> Unit)? = null,
    onTap: ((T) -> (() -> Unit)?)? = null,
    onLongPress: ((T) -> (() -> Unit)?)? = null,
    cardPadding: PaddingValues = kGroupedCardPadding,
    backgroundBuilder: ((T) -> (@Composable () -> Unit)?)? = null,
    cardGesture: (@Composable (T) -> Modifier)? = null,
    itemBuilder: @Composable (T) -> Unit,
) {
    val last = items.size - 1
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = kGroupedHorizontalPadding),
    ) {
        if (header != null || title != null) GroupTitle(title, header)
        items.forEachIndexed { index, item ->
            if (index > 0) Spacer(Modifier.height(kGroupedGap))
            GroupedCard(
                shape = listShape(index, last),
                onTap = onTap?.invoke(item),
                onLongPress = onLongPress?.invoke(item),
                padding = cardPadding,
                background = backgroundBuilder?.invoke(item),
                modifier = Modifier.fillMaxWidth(),
                gestureModifier = cardGesture?.invoke(item) ?: Modifier,
            ) {
                itemBuilder(item)
            }
        }
    }
}

@Composable
fun <T> GroupedCardGrid(
    items: List<T>,
    title: String? = null,
    header: (@Composable () -> Unit)? = null,
    onTap: ((T) -> (() -> Unit)?)? = null,
    backgroundBuilder: ((T) -> (@Composable () -> Unit)?)? = null,
    cardPadding: PaddingValues = kGroupedCardPadding,
    maxCrossAxisExtent: Dp = kGroupedGridMaxExtent,
    mainAxisExtent: Dp? = kGroupedGridTileHeight,
    crossAxisCount: Int? = null,
    spacing: Dp = kGroupedGap,
    itemBuilder: @Composable (T) -> Unit,
) {
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = kGroupedHorizontalPadding),
    ) {
        if (header != null || title != null) GroupTitle(title, header)
        androidx.compose.foundation.layout.BoxWithConstraints(Modifier.fillMaxWidth()) {
            val width = maxWidth
            val columns = crossAxisCount
                ?: max(1, ceil((width / (maxCrossAxisExtent + spacing)).toDouble()).toInt())

            Column(Modifier.fillMaxWidth()) {
                var start = 0
                var rowIndex = 0
                while (start < items.size) {
                    if (rowIndex > 0) Spacer(Modifier.height(spacing))
                    val rowModifier =
                        if (mainAxisExtent == null) {
                            Modifier.fillMaxWidth().height(IntrinsicSize.Max)
                        } else {
                            Modifier.fillMaxWidth()
                        }
                    Row(rowModifier, verticalAlignment = Alignment.CenterVertically) {
                        for (c in 0 until columns) {
                            if (c > 0) Spacer(Modifier.width(spacing))
                            val index = start + c
                            Box(Modifier.weight(1f)) {
                                if (index < items.size) {
                                    val shape = gridShape(index, columns, items.size)
                                    val cellModifier = if (mainAxisExtent != null) {
                                        Modifier.fillMaxWidth().height(mainAxisExtent)
                                    } else {
                                        Modifier.fillMaxWidth()
                                    }
                                    GroupedCard(
                                        shape = shape,
                                        onTap = onTap?.invoke(items[index]),
                                        padding = cardPadding,
                                        background = backgroundBuilder?.invoke(items[index]),
                                        contentAlignment = if (mainAxisExtent == null) {
                                            Alignment.TopStart
                                        } else {
                                            Alignment.CenterStart
                                        },
                                        modifier = cellModifier,
                                    ) {
                                        itemBuilder(items[index])
                                    }
                                }
                            }
                        }
                    }
                    start += columns
                    rowIndex++
                }
            }
        }
    }
}

private fun gridShape(index: Int, columns: Int, total: Int): RoundedCornerShape {
    val col = index % columns
    val top = index - columns < 0
    val bottom = index + columns >= total
    val left = col == 0
    val right = col == columns - 1 || index + 1 >= total
    fun corner(a: Boolean, b: Boolean): Dp =
        if (a && b) kGroupedOuterRadius else kGroupedInnerRadius
    return RoundedCornerShape(
        topStart = corner(top, left),
        topEnd = corner(top, right),
        bottomStart = corner(bottom, left),
        bottomEnd = corner(bottom, right),
    )
}
